app/src/main/java/com/github/studydistractor/sdp/eventHistory/EventHistoryServiceFirebase.kt

```diff
@@ -0,0 +1,107 @@
+import com.google.android.gms.tasks.Task
+import com.google.android.gms.tasks.Tasks
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.DatabaseReference
+import com.google.firebase.database.FirebaseDatabase
+import com.google.firebase.database.ValueEventListener
+
+class EventHistoryServiceFirebase: EventHistoryModel {
+    private val eventPaths = "Events"
+    private val eventDatabaseRef: DatabaseReference = FirebaseDatabase.getInstance().getReference(eventPaths)
+    private val eventParticipantsPaths = "EventParticipants"
+    private val eventParticipantsDatabaseRef: DatabaseReference = FirebaseDatabase.getInstance().getReference(eventParticipantsPaths)
+    private var auth : FirebaseAuth = FirebaseAuth.getInstance()
+
+    val firebaseEvents = mutableStateListOf<FirebaseEvent>()
+    val firebaseEventParticipants = mutableStateListOf<FirebaseEventParticipants>()
+
+    init {
+        eventDatabaseRef.addListenerForSingleValueEvent(object : ValueEventListener {
+            override fun onDataChange(snapshot: DataSnapshot) {
+                firebaseEvents.clear()
+                for (event in snapshot.children) {
+                    val eventItem = event.getValue(FirebaseEvent::class.java)
+                    if (eventItem != null) {
+                        firebaseEvents.add(eventItem)
+                    }
+                }
+            }
+
+            override fun onCancelled(error: DatabaseError) {
+                Log.d("Firebase Event", "loadPost:onCancelled " + error.toException().toString())
+            }
+        })
+
+        eventParticipantsDatabaseRef.addListenerForSingleValueEvent(object : ValueEventListener {
+            val eventParticipants = mutableStateListOf<FirebaseEventParticipants>()
+            override fun onDataChange(snapshot: DataSnapshot) {
+                firebaseEventParticipants.clear()
+                for (eP in snapshot.children) {
+                    val eventParticipantsItem = eP.getValue(FirebaseEventParticipants::class.java)
+                    if (eventParticipantsItem != null) {
+                        firebaseEventParticipants.add(eventParticipantsItem)
+                    }
+                }
+            }
+
+            override fun onCancelled(error: DatabaseError) {
+                Log.d("Firebase Event Participant", "loadPost:onCancelled " + error.toException().toString())
+            }
+        })
+    }
+
+    override fun getEvents(): Task<List<Event>>{
+        val events = mutableListOf<Event>()
+
+        for (fe in firebaseEvents){
+            val ev : Event
+            try {
+                ev = fe.toEvent()
+            } catch(e: Exception){
+                continue
+            }
+
+            events.add(ev)
+        }
+
+        return Tasks.forResult(events)
```

Comment:
Not the good way to do it. It will not be update in direct. You should use a observer to see update.

Reply:
> Not the good way to do it. It will not be update in direct. You should use a observer to see update.

Thanks for your tips ! Now it works and it is clean !